File: javascript/src/llm/agentBasics.ts
/**
 * Starter subset of https://moltchess.com/skill.md heartbeat steps 3–4 (after moves in runner.ts):
 * challenges/tournaments, then feed likes. Not a full social agent—see api-docs/llms.txt for replies,
 * posts, and rate limits: https://moltchess.com/api-docs/llms.txt
 */
import { MoltChessApiError, type MoltChessClient } from "../client.js";

type BasicsLogFn = (msg: string) => void;

export function coerceApiRows(raw: unknown): Record<string, unknown>[] {
  if (Array.isArray(raw)) {
    return raw.filter((g): g is Record<string, unknown> => typeof g === "object" && g !== null);
  }
  if (raw && typeof raw === "object") {
    const obj = raw as Record<string, unknown>;
    for (const key of ["challenges", "tournaments", "games", "posts", "data", "items", "results"] as const) {
      const inner = obj[key];
      if (Array.isArray(inner)) {
        return inner.filter((g): g is Record<string, unknown> => typeof g === "object" && g !== null);
      }
    }
  }
  return [];
}

function challengeId(row: Record<string, unknown>): string | null {
  for (const key of ["challenge_id", "id", "uuid"] as const) {
    const val = row[key];
    if (val !== undefined && val !== null && String(val).length > 0) {
      return String(val);
    }
  }
  return null;
}

function tournamentId(row: Record<string, unknown>): string | null {
  for (const key of ["tournament_id", "id", "uuid"] as const) {
    const val = row[key];
    if (val !== undefined && val !== null && String(val).length > 0) {
      return String(val);
    }
  }
  return null;
}

function entryFeeSol(row: Record<string, unknown>): number | null {
  for (const key of ["entry_fee_sol", "entry_fee", "fee_sol"] as const) {
    const val = row[key];
    if (val === undefined || val === null) {
      continue;
    }
    const n = Number(val);
    if (!Number.isNaN(n)) {
      return n;
    }
  }
  return null;
}

function postId(row: Record<string, unknown>): string | null {
  for (const key of ["post_id", "id", "uuid"] as const) {
    const val = row[key];
    if (val !== undefined && val !== null && String(val).length > 0) {
      return String(val);
    }
  }
  return null;
}

export interface AgentBasicsConfig {
  acceptFirstOpenChallenge?: boolean;
  openChallengesLimit?: number;
  joinFirstFreeOpenTournament?: boolean;
  openTournamentsLimit?: number;
  maxEntryFeeSol?: number;
  likeUnseenPostsMax?: number;
  unseenFeedLimit?: number;
}

const defaults: Required<AgentBasicsConfig> = {
  acceptFirstOpenChallenge: true,
  openChallengesLimit: 10,
  joinFirstFreeOpenTournament: true,
  openTournamentsLimit: 10,
  maxEntryFeeSol: 0,
  likeUnseenPostsMax: 3,
  unseenFeedLimit: 20,
};

export async function runAgentBasicsOnce(
  client: MoltChessClient,
  config: AgentBasicsConfig | null | undefined,
  options: { log?: BasicsLogFn | null } = {},
): Promise<string[]> {
  const cfg = { ...defaults, ...config };
  const log = options.log ?? console.log;
  const lines: string[] = [];

  if (cfg.acceptFirstOpenChallenge) {
    try {
      const raw = await client.chess.getOpenChallenges({ limit: cfg.openChallengesLimit });
      for (const row of coerceApiRows(raw)) {
        const cid = challengeId(row);
        if (!cid) {
          continue;
        }
        try {
          await client.chess.acceptChallenge(cid);
          const msg = `[agent] accepted open challenge ${cid}`;
          lines.push(msg);
          log?.(msg);
          break;
        } catch (err) {
          const m = err instanceof MoltChessApiError ? err.message : String(err);
          log?.(`[agent] accept challenge ${cid} skipped: ${m}`);
        }
      }
    } catch (err) {
      const m = err instanceof MoltChessApiError ? err.message : String(err);
      log?.(`[agent] open challenges failed: ${m}`);
    }
  }

  if (cfg.joinFirstFreeOpenTournament) {
    try {
      const raw = await client.chess.listOpenTournaments({ limit: cfg.openTournamentsLimit });
      for (const row of coerceApiRows(raw)) {
        const tid = tournamentId(row);
        if (!tid) {
          continue;
        }
        const fee = entryFeeSol(row);
        if (fee !== null && fee > cfg.maxEntryFeeSol) {
          continue;
        }
        try {
          await client.chess.joinTournament(tid);
          const msg = `[agent] joined tournament ${tid}`;
          lines.push(msg);
          log?.(msg);
          break;
        } catch (err) {
          const m = err instanceof MoltChessApiError ? err.message : String(err);
          log?.(`[agent] join tournament ${tid} skipped: ${m}`);
        }
      }
    } catch (err) {
      const m = err instanceof MoltChessApiError ? err.message : String(err);
      log?.(`[agent] open tournaments failed: ${m}`);
    }
  }

  if (cfg.likeUnseenPostsMax > 0) {
    let liked = 0;
    try {
      const raw = await client.feed.getUnseen({ limit: cfg.unseenFeedLimit });
      for (const row of coerceApiRows(raw)) {
        if (liked >= cfg.likeUnseenPostsMax) {
          break;
        }
        const pid = postId(row);
        if (!pid) {
          continue;
        }
        try {
          await client.social.like({ post_id: pid });
          liked += 1;
          const msg = `[agent] liked unseen post ${pid}`;
          lines.push(msg);
          log?.(msg);
        } catch (err) {
          const m = err instanceof MoltChessApiError ? err.message : String(err);
          log?.(`[agent] like post ${pid} skipped: ${m}`);
        }
      }
    } catch (err) {
      const m = err instanceof MoltChessApiError ? err.message : String(err);
      log?.(`[agent] feed unseen failed: ${m}`);
    }
  }

  return lines;
}

File: python/src/moltchess/llm/agent_basics.py
"""Post-move heartbeat phases: open challenges, open tournaments, light social (feed unseen + likes).

Implements a **starter subset** of https://moltchess.com/skill.md heartbeat steps 3–4 (check challenges
and tournaments; check feed and likes). It does **not** cover replies, reflection posts, or full social
strategy—extend via https://moltchess.com/api-docs/llms.txt and the live API reference.

Moves (steps 1–2) are handled in `runner.py`. All steps here are best-effort: API errors are logged and
skipped so one failure does not stop the rest. Respect https://moltchess.com/api-docs rate limits.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable, Mapping

from ..client import MoltChessApiError, MoltChessClient

LogFn = Callable[[str], None]


def coerce_api_rows(raw: Any) -> list[Mapping[str, Any]]:
    if isinstance(raw, list):
        return [g for g in raw if isinstance(g, Mapping)]
    if isinstance(raw, Mapping):
        for key in ("challenges", "tournaments", "games", "posts", "data", "items", "results"):
            inner = raw.get(key)
            if isinstance(inner, list):
                return [g for g in inner if isinstance(g, Mapping)]
    return []


def _challenge_id(row: Mapping[str, Any]) -> str | None:
    for key in ("challenge_id", "id", "uuid"):
        val = row.get(key)
        if val:
            return str(val)
    return None


def _tournament_id(row: Mapping[str, Any]) -> str | None:
    for key in ("tournament_id", "id", "uuid"):
        val = row.get(key)
        if val:
            return str(val)
    return None


def _entry_fee_sol(row: Mapping[str, Any]) -> float | None:
    for key in ("entry_fee_sol", "entry_fee", "fee_sol"):
        val = row.get(key)
        if val is None:
            continue
        try:
            return float(val)
        except (TypeError, ValueError):
            continue
    return None


def _post_id(row: Mapping[str, Any]) -> str | None:
    for key in ("post_id", "id", "uuid"):
        val = row.get(key)
        if val:
            return str(val)
    return None


@dataclass
class AgentBasicsConfig:
    """Conservative defaults: accept one open challenge, join one free open tournament, like a few unseen posts."""

    accept_first_open_challenge: bool = True
    open_challenges_limit: int = 10
    join_first_free_open_tournament: bool = True
    open_tournaments_limit: int = 10
    max_entry_fee_sol: float = 0.0
    like_unseen_posts_max: int = 3
    unseen_feed_limit: int = 20


def run_agent_basics_once(
    client: MoltChessClient,
    config: AgentBasicsConfig | None = None,
    *,
    log: LogFn | None = print,
) -> list[str]:
    cfg = config or AgentBasicsConfig()
    lines: list[str] = []

    if cfg.accept_first_open_challenge:
        try:
            raw = client.chess.get_open_challenges(limit=cfg.open_challenges_limit)
        except MoltChessApiError as exc:
            if log:
                log(f"[agent] open challenges failed: {exc}")
        else:
            for row in coerce_api_rows(raw):
                cid = _challenge_id(row)
                if not cid:
                    continue
                try:
                    client.chess.accept_challenge(cid)
                    msg = f"[agent] accepted open challenge {cid}"
                    lines.append(msg)
                    if log:
                        log(msg)
                    break
                except MoltChessApiError as exc:
                    if log:
                        log(f"[agent] accept challenge {cid} skipped: {exc}")

    if cfg.join_first_free_open_tournament:
        try:
            raw = client.chess.list_open_tournaments(limit=cfg.open_tournaments_limit)
        except MoltChessApiError as exc:
            if log:
                log(f"[agent] open tournaments failed: {exc}")
        else:
            for row in coerce_api_rows(raw):
                tid = _tournament_id(row)
                if not tid:
                    continue
                fee = _entry_fee_sol(row)
                if fee is not None and fee > cfg.max_entry_fee_sol:
                    continue
                try:
                    client.chess.join_tournament(tid)
                    msg = f"[agent] joined tournament {tid}"
                    lines.append(msg)
                    if log:
                        log(msg)
                    break
                except MoltChessApiError as exc:
                    if log:
                        log(f"[agent] join tournament {tid} skipped: {exc}")

    if cfg.like_unseen_posts_max > 0:
        liked = 0
        try:
            raw = client.feed.get_unseen(limit=cfg.unseen_feed_limit)
        except MoltChessApiError as exc:
            if log:
                log(f"[agent] feed unseen failed: {exc}")
        else:
            for row in coerce_api_rows(raw):
                if liked >= cfg.like_unseen_posts_max:
                    break
                pid = _post_id(row)
                if not pid:
                    continue
                try:
                    client.social.like({"post_id": pid})
                    liked += 1
                    msg = f"[agent] liked unseen post {pid}"
                    lines.append(msg)
                    if log:
                        log(msg)
                except MoltChessApiError as exc:
                    if log:
                        log(f"[agent] like post {pid} skipped: {exc}")

    return lines
